hypothesis/lms
# New unsubscribe mechanism for digest emails

Adds the basic model, service and view to handle email unsubscribed.


Needs: https://github.com/hypothesis/lms/pull/5261

## Before merging:

- [x] Set the new en var in all environments

## After merging 

- [ ] Update the digest template to include the link
- [ ] Remove rule to include mailchimp own link

## Testing

- Tweak the sending code to use a new test template:

```diff 
--git a/lms/services/digest.py b/lms/services/digest.py
index 37c4b6f69..a87fdffd5 100644
--- a/lms/services/digest.py
+++ b/lms/services/digest.py
@@ -72,7 +72,8 @@ class DigestService:
                 continue
 
             self._mailchimp_service.send_template(
-                "instructor-email-digest",
+                # "instructor-email-digest",
+                "test-unsubcribe-link",
                 self._sender,
                 recipient=EmailRecipient(to_email, unified_user.display_name),
                 template_vars=digest,
```


- Check the new template 

https://mandrillapp.com/templates/code?id=test-unsubcribe-link

The HTML is copied from GH simple emails and the wording from mailchimp unsubscribe links.



- Log in to https://hypothesis.instructure.com/ first as `eng+canvasteacher@hypothes.is` and then as `eng+canvasteacher2@hypothes.is` and launch [localhost (make devdata) HTML Assignment](https://hypothesis.instructure.com/courses/125/assignments/873) as each. This is to create the records of the instructors in your local DB.

- Log in as `eng+canvasstudent@hypothes.is`, launch the same assignment, and create an annotation.


- Go over http://localhost:8001/admin/email and kick-off a task

Include the ids: 

```
acct:3a022b6c146dfd9df4ea8662178eac@lms.hypothes.is acct:9ea84c4bcfd9329b4495f5b27d8cb6@lms.hypothes.is
```

and the relevant dates.


- Check the values in the console, a new unsubscribe variable will appear now and also the headers dict will be present.

- See the "sent" emails over: https://mandrillapp.com/activity, click the link there in one of them (you might have to copied it, at least in chrome)

- You'll see a page with `You've been unsubscribed` 

- There's a new row reflecting that on the DB:

```
tox -qe dockercompose -- exec postgres psql -U postgres -c "select * from email_unsubscribe";
          created           |          updated           | id |        tag        |                      h_userid                       
----------------------------+----------------------------+----+-------------------+-----------------------------------------------------
 2023-04-13 08:25:30.864655 | 2023-04-13 08:25:30.864655 |  3 | instructor_digest | acct:9ea84c4bcfd9329b4495f5b27d8cb6@lms.hypothes.is       
```


- Back on  http://localhost:8001/admin/email, try to send the email again, it should only send one email now.


- Remove the unsubscribe record: 

```
tox -qe dockercompose -- exec postgres psql -U postgres -c "truncate email_unsubscribe";
```


- Send again on  http://localhost:8001/admin/email. It should send both it this time.
